src/rules/no-duplicate-headings.js

```diff
@@ -29,46 +30,91 @@ export default {
 		messages: {
 			duplicateHeading: 'Duplicate heading "{{text}}" found.',
 		},
+
+		schema: [
+			{
+				type: "object",
+				properties: {
+					checkSiblingsOnly: {
+						type: "boolean",
+					},
+				},
+				additionalProperties: false,
+			},
+		],
+
+		defaultOptions: [{ checkSiblingsOnly: false }],
 	},
 
 	create(context) {
-		const headings = new Set();
+		const [{ checkSiblingsOnly }] = context.options;
 		const { sourceCode } = context;
 
+		const headingsByLevel = checkSiblingsOnly
+			? new Array(7).fill(null).map(() => new Set())
+			: [null, new Set()];
+		let lastLevel = 1;
+		let currentLevelHeadings = headingsByLevel[lastLevel];
+
+		/**
+		 * Gets the text of a heading node
+		 * @param {HeadingNode} node The heading node
+		 * @returns {string} The heading text
+		 */
+		function getHeadingText(node) {
+			/*
+			 * There are two types of headings in markdown:
+			 * - ATX headings, which start with one or more # characters
+			 * - Setext headings, which are underlined with = or -
+			 * Setext headings are identified by being on two lines instead of one,
+			 * with the second line containing only = or - characters. In order to
+			 * get the correct heading text, we need to determine which type of
+			 * heading we're dealing with.
+			 */
+			const isSetext =
+				node.position.start.line !== node.position.end.line;
+
+			return isSetext
+				? // get only the text from the first line
+					sourceCode.lines[node.position.start.line - 1].trim()
+				: // get the text without the leading # characters
+					sourceCode
+						.getText(node)
+						.slice(node.depth + 1)
+						.trim();
+		}
+
 		return {
 			heading(node) {
-				/*
-				 * There are two types of headings in markdown:
-				 * - ATX headings, which start with one or more # characters
-				 * - Setext headings, which are underlined with = or -
-				 * Setext headings are identified by being on two lines instead of one,
-				 * with the second line containing only = or - characters. In order to
-				 * get the correct heading text, we need to determine which type of
-				 * heading we're dealing with.
-				 */
-				const isSetext =
-					node.position.start.line !== node.position.end.line;
-
-				const text = isSetext
-					? // get only the text from the first line
-						sourceCode.lines[node.position.start.line - 1].trim()
-					: // get the text without the leading # characters
-						sourceCode
-							.getText(node)
-							.slice(node.depth + 1)
-							.trim();
-
-				if (headings.has(text)) {
+				const headingText = getHeadingText(node);
+
+				if (checkSiblingsOnly) {
+					const currentLevel = node.depth;
+
+					const direction = currentLevel > lastLevel ? 1 : -1;
+					while (lastLevel !== currentLevel) {
+						if (direction > 0) {
```

Comment:
```suggestion
						if (direction > 0) {
```

Like the above comment, could you clarify the conditional statement?

Maybe we can use `if (direction === UP)` or something similar.